File: consorcio/lib/infrastructure/firestore/firebase_attendance_permission_request_repository.dart
import 'dart:async';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_storage/firebase_storage.dart';

import '../../domain/entities/attendance_permission_request.dart';
import '../../domain/errors/domain_exception.dart';
import '../../domain/repositories/attendance_permission_request_repository.dart';
import '../../domain/repositories/folder_image_repository.dart';
import 'firestore_client.dart';

class FirebaseAttendancePermissionRequestRepository
    implements AttendancePermissionRequestRepository {
  FirebaseAttendancePermissionRequestRepository({
    FirebaseFirestore? firestore,
    FirebaseStorage? storage,
  })  : _firestore = firestore ?? FirebaseFirestore.instance,
        _storage = storage ?? FirebaseStorage.instance;

  final FirebaseFirestore _firestore;
  final FirebaseStorage _storage;

  CollectionReference<Map<String, dynamic>> get _requests =>
      _firestore.collection('attendancePermissionRequests');

  AttendancePermissionRequest _map(String id, Map<String, dynamic> data) {
    return AttendancePermissionRequest(
      id: id,
      userId: data['userId'] as String? ?? '',
      userName: data['userName'] as String? ?? '',
      description: data['description'] as String? ?? '',
      status: data['status'] as String? ??
          AttendancePermissionRequestStatus.pending,
      evidencePhotoUrl: data['evidencePhotoUrl'] as String?,
      leaveDays: (data['leaveDays'] as num?)?.toInt(),
      startDateKey: data['startDateKey'] as String?,
      endDateKey: data['endDateKey'] as String?,
      adminNote: data['adminNote'] as String?,
      createdAt:
          (data['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
    );
  }

  @override
  Future<AttendancePermissionRequest> create({
    required String userId,
    required String userName,
    required String description,
    ImageFilePayload? evidencePhoto,
  }) async {
    String? photoUrl;
    String? storagePath;
    if (evidencePhoto != null) {
      final fileId = DateTime.now().millisecondsSinceEpoch.toString();
      storagePath = 'attendances/$userId/permission-requests/$fileId.jpg';
      try {
        final ref = _storage.ref(storagePath);
        await ref
            .putData(
              evidencePhoto.bytes,
              SettableMetadata(
                contentType: evidencePhoto.contentType.isEmpty
                    ? 'image/jpeg'
                    : evidencePhoto.contentType,
              ),
            )
            .timeout(const Duration(seconds: 20));
        photoUrl = await ref.getDownloadURL().timeout(
          const Duration(seconds: 10),
        );
      } on DomainException {
        rethrow;
      } catch (_) {
        throw DomainException(
          'No se pudo subir la foto. Revisa tu conexión.',
        );
      }
    }

    final payload = <String, dynamic>{
      'userId': userId,
      'userName': userName.trim().isEmpty ? 'Usuario' : userName.trim(),
      'description': description,
      'status': AttendancePermissionRequestStatus.pending,
      'createdAt': Timestamp.now(),
    };
    if (photoUrl != null && storagePath != null) {
      payload['evidencePhotoUrl'] = photoUrl;
      payload['evidencePhotoPath'] = storagePath;
    }

    try {
      final ref = _requests.doc();
      await setQueued(ref, payload, timeout: const Duration(seconds: 8));
      return _map(ref.id, {
        ...payload,
        'createdAt': payload['createdAt'],
      });
    } on FirebaseException catch (error) {
      if (error.code == 'permission-denied') {
        throw DomainException(
          'No se pudo enviar la solicitud (permiso denegado). '
          'Cierra sesión, vuelve a entrar e intenta de nuevo.',
        );
      }
      if (error.code == 'unavailable' || error.code == 'network-request-failed') {
        throw DomainException(
          'Sin conexión. Revisa tu internet e intenta de nuevo.',
        );
      }
      throw DomainException(
        'No se pudo enviar la solicitud de permiso (${error.code}).',
      );
    }
  }

  @override
  Future<List<AttendancePermissionRequest>> listByUser(String userId) async {
    try {
      final snapshot = await queryFast(
        _requests
            .where('userId', isEqualTo: userId)
            .orderBy('createdAt', descending: true)
            .limit(20),
        timeout: const Duration(seconds: 8),
      );
      return snapshot.docs.map((doc) => _map(doc.id, doc.data())).toList();
    } catch (_) {
      return const [];
    }
  }

  @override
  Future<List<AttendancePermissionRequest>> listPendingByUser(
    String userId,
  ) async {
    try {
      final snapshot = await _requests
          .where('userId', isEqualTo: userId)
          .where(
            'status',
            isEqualTo: AttendancePermissionRequestStatus.pending,
          )
          .limit(1)
          .get()
          .timeout(const Duration(seconds: 6));
      return snapshot.docs.map((doc) => _map(doc.id, doc.data())).toList();
    } on FirebaseException catch (error) {
      if (error.code == 'failed-precondition' ||
          error.code == 'permission-denied') {
        return const [];
      }
      rethrow;
    } on TimeoutException {
      return const [];
    }
  }
}

File: src/infrastructure/firestore/FirebaseAttendancePermissionRequestRepository.ts
import { FirebaseError } from 'firebase/app'
import {
  addDoc,
  collection,
  doc,
  getDoc,
  getDocs,
  orderBy,
  query,
  Timestamp,
  updateDoc,
  where,
} from 'firebase/firestore'
import {
  getDownloadURL,
  ref,
  uploadBytes,
} from 'firebase/storage'
import type { AttendancePermissionRequest } from '@/domain/entities/AttendancePermissionRequest'
import {
  AttendancePermissionRequestStatus,
  isAttendancePermissionRequestStatus,
} from '@/domain/entities/AttendancePermissionRequest'
import type {
  ApproveAttendancePermissionRequestInput,
  AttendancePermissionRequestRepository,
  CreateAttendancePermissionRequestInput,
  RejectAttendancePermissionRequestInput,
  UploadPermissionEvidenceInput,
} from '@/domain/repositories/AttendancePermissionRequestRepository'
import { UnauthorizedError, ValidationError } from '@/domain/errors/DomainError'
import { firestoreDb, firebaseStorage } from '@/infrastructure/firebase/firebaseApp'

interface PermissionRequestDoc {
  userId: string
  userName: string
  description: string
  status: string
  evidencePhotoUrl?: string
  evidencePhotoPath?: string
  leaveDays?: number
  startDateKey?: string
  endDateKey?: string
  adminNote?: string
  reviewedById?: string
  reviewedByName?: string
  reviewedAt?: Timestamp
  createdAt: Timestamp
}

function mapRequest(
  id: string,
  data: PermissionRequestDoc,
): AttendancePermissionRequest | null {
  if (!isAttendancePermissionRequestStatus(data.status)) return null
  return {
    id,
    userId: data.userId,
    userName: data.userName,
    description: data.description,
    status: data.status,
    evidencePhotoUrl: data.evidencePhotoUrl || undefined,
    evidencePhotoPath: data.evidencePhotoPath || undefined,
    leaveDays:
      typeof data.leaveDays === 'number' ? data.leaveDays : undefined,
    startDateKey: data.startDateKey || undefined,
    endDateKey: data.endDateKey || undefined,
    adminNote: data.adminNote || undefined,
    reviewedById: data.reviewedById || undefined,
    reviewedByName: data.reviewedByName || undefined,
    reviewedAt: data.reviewedAt?.toDate?.(),
    createdAt: data.createdAt?.toDate?.() ?? new Date(),
  }
}

function mapWriteError(error: unknown, fallback: string): never {
  if (error instanceof ValidationError || error instanceof UnauthorizedError) {
    throw error
  }
  if (error instanceof FirebaseError && error.code === 'permission-denied') {
    throw new UnauthorizedError(
      'No se pudo guardar la solicitud de permiso. Revisa tu sesión.',
    )
  }
  throw new ValidationError(fallback)
}

export class FirebaseAttendancePermissionRequestRepository
  implements AttendancePermissionRequestRepository
{
  private readonly collectionRef = collection(
    firestoreDb,
    'attendancePermissionRequests',
  )

  async create(
    input: CreateAttendancePermissionRequestInput,
  ): Promise<AttendancePermissionRequest> {
    const now = Timestamp.now()
    const payload: PermissionRequestDoc = {
      userId: input.userId,
      userName: input.userName,
      description: input.description,
      status: AttendancePermissionRequestStatus.Pending,
      createdAt: now,
    }
    if (input.evidencePhotoUrl && input.evidencePhotoPath) {
      payload.evidencePhotoUrl = input.evidencePhotoUrl
      payload.evidencePhotoPath = input.evidencePhotoPath
    }
    try {
      const created = await addDoc(this.collectionRef, payload)
      return mapRequest(created.id, payload) as AttendancePermissionRequest
    } catch (error) {
      mapWriteError(error, 'No se pudo enviar la solicitud de permiso')
    }
  }

  async getById(requestId: string): Promise<AttendancePermissionRequest | null> {
    const snapshot = await getDoc(doc(this.collectionRef, requestId))
    if (!snapshot.exists()) return null
    return mapRequest(snapshot.id, snapshot.data() as PermissionRequestDoc)
  }

  async listPending(): Promise<AttendancePermissionRequest[]> {
    const snapshot = await getDocs(
      query(
        this.collectionRef,
        where('status', '==', AttendancePermissionRequestStatus.Pending),
        orderBy('createdAt', 'asc'),
      ),
    )
    return snapshot.docs
      .map((item) => mapRequest(item.id, item.data() as PermissionRequestDoc))
      .filter((item): item is AttendancePermissionRequest => item !== null)
  }

  async listByUser(userId: string): Promise<AttendancePermissionRequest[]> {
    const snapshot = await getDocs(
      query(
        this.collectionRef,
        where('userId', '==', userId),
        orderBy('createdAt', 'desc'),
      ),
    )
    return snapshot.docs
      .map((item) => mapRequest(item.id, item.data() as PermissionRequestDoc))
      .filter((item): item is AttendancePermissionRequest => item !== null)
  }

  async listPendingByUser(
    userId: string,
  ): Promise<AttendancePermissionRequest[]> {
    const snapshot = await getDocs(
      query(
        this.collectionRef,
        where('userId', '==', userId),
        where('status', '==', AttendancePermissionRequestStatus.Pending),
      ),
    )
    return snapshot.docs
      .map((item) => mapRequest(item.id, item.data() as PermissionRequestDoc))
      .filter((item): item is AttendancePermissionRequest => item !== null)
  }

  async approve(
    input: ApproveAttendancePermissionRequestInput,
  ): Promise<AttendancePermissionRequest> {
    const refDoc = doc(this.collectionRef, input.requestId)
    const now = Timestamp.now()
    try {
      await updateDoc(refDoc, {
        status: AttendancePermissionRequestStatus.Approved,
        leaveDays: input.leaveDays,
        startDateKey: input.startDateKey,
        endDateKey: input.endDateKey,
        adminNote: input.adminNote ?? '',
        reviewedById: input.reviewedById,
        reviewedByName: input.reviewedByName,
        reviewedAt: now,
      })
    } catch (error) {
      mapWriteError(error, 'No se pudo aprobar la solicitud')
    }
    const updated = await this.getById(input.requestId)
    if (!updated) throw new ValidationError('No se pudo aprobar la solicitud')
    return updated
  }

  async reject(
    input: RejectAttendancePermissionRequestInput,
  ): Promise<AttendancePermissionRequest> {
    const refDoc = doc(this.collectionRef, input.requestId)
    const now = Timestamp.now()
    try {
      await updateDoc(refDoc, {
        status: AttendancePermissionRequestStatus.Rejected,
        adminNote: input.adminNote ?? '',
        reviewedById: input.reviewedById,
        reviewedByName: input.reviewedByName,
        reviewedAt: now,
      })
    } catch (error) {
      mapWriteError(error, 'No se pudo rechazar la solicitud')
    }
    const updated = await this.getById(input.requestId)
    if (!updated) throw new ValidationError('No se pudo rechazar la solicitud')
    return updated
  }

  async uploadEvidencePhoto(
    input: UploadPermissionEvidenceInput,
  ): Promise<{ url: string; path: string }> {
    const path = `attendances/${input.userId}/permission-requests/${input.fileId}.jpg`
    const storageRef = ref(firebaseStorage, path)
    try {
      await uploadBytes(storageRef, input.data, {
        contentType: input.contentType || 'image/jpeg',
      })
      const url = await getDownloadURL(storageRef)
      return { url, path }
    } catch (error) {
      mapWriteError(error, 'No se pudo subir la foto de la solicitud')
    }
  }
}
